store/iterator.go

```diff
@@ -9,40 +9,50 @@ import (
 	"github.com/iov-one/weave/errors"
 )
 
-///////////////////////////////////////////////////////
-// From Items to Iterator
-
 type btreeIter struct {
-	read      <-chan btree.Item
-	stop      chan<- struct{}
-	onceStop  sync.Once
-	onceRead  sync.Once
+	read <-chan btree.Item
+
+	// Stop is used to signal that the iterator should stop and free
+	// resources.
+	stop     chan<- struct{}
+	onceStop sync.Once
+
+	// Released is used to signal that the iterator was closed and all
+	// resources are free.
+	released <-chan struct{}
+
 	ascending bool
 }
 
 // combine joins our results with those of the parent,
 // taking into consideration overwrites and deletes...
 func ascendBtree(bt *btree.BTree, start, end []byte) *btreeIter {
 	read := make(chan btree.Item)
-	// ensure we never block when we call close()
-	stop := make(chan struct{}, 1)
+	stop := make(chan struct{})
+	released := make(chan struct{})
+
 	iter := &btreeIter{
 		read:      read,
 		stop:      stop,
+		released:  released,
 		ascending: true,
 	}
 
-	insert := func(item btree.Item) bool {
-		select {
-		case read <- item:
-			return true
-		case <-stop:
-			iter.onceRead.Do(func() { close(read) })
-			return false
+	go func() {
+		defer func() {
+			close(read)
+			close(released)
+		}()
+
+		insert := func(item btree.Item) bool {
```

Comment:
Why do you define `insert` inside the goroutine? It is just a callback, and seems clearer if this is outside the go-routine

Reply:
I think it is easier to read. It helped me once I understood the functionality.

I love to read top-down, with as little helper functions as possible. If not that `insert` is quite some lines, I would inline the code instead. In my opinion `insert` belongs to this gorutine as it must be used only inside of it. 

I am looking at [the code](https://github.com/iov-one/weave/blob/341fb17e459d368049a441733d651e5afb188a32/store/iterator.go#L29-L67) again and I think the best desciption of my intention is to put `inser` as close to the use of it as possible.

Does this make sense? I can see that you have the opposite preference and for sure you have a good reason for this. Can you maybe explain me what is your motivation?

Reply:
I would normally agree, but since the function is pretty small I think it reads cleaner to define this outside the goroutine as the function does not use any of the variables declared inside of it.

Reply:
Fair enough.

It is just personal preference from the two of us. In a draw, PR author wins :smile: 

Merge as is... there will be another update for context, and hopefully we get a third opinion then. For now, just merge this fix. Good to find it and improve the code.